src/group_theory/free_product.lean

```diff
@@ -0,0 +1,197 @@
+/-
+Copyright (c) 2021 David Wärn. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Wärn
+-/
+import data.list.chain
+/-!
+# The free product of groups or monoids
+
+Given an `ι`-indexed family `M` of monoids, where both `ι` and all `M i` have decidable
+equality, we define the free product `free_product M` of `M` as the type of reduced words, and show
+that it has the universal property of a coproduct. One could define the free product more abstractly
+in terms of its presentation, but this definition has the advantage of giving decidable equality.
+
+When `M i` are all groups, we show that `free_product M` is also a group.
+
+## Main definitions
+
+- `free_product M`
+- `of {i}`: the embedding `M i →* free_product M`.
+- `lift X : (Π {i}, M i →* X) ≃ (free_product M →* X)`: the free product is a categorical coproduct.
+
+## Idea
+
+We define the monoid structure on `free_product M` by letting each `M i` act on `free_product M`.
+One way to understand this is that each `M i` thus maps into the monoid
+`free_product M → free_product M`, and the submonoid generated by their images bijects with
+`free_product M` by evaluation at `[]`, inducing a monoid structure on `free_product M`.
+We take the shortcut of not introducing `free_product M → free_product M` or this bijection.
+
+## References
+
+[van der Waerden, *Free products of groups*][MR25465]
+
+-/
+
+variables {ι : Type*} (M : Π i : ι, Type*) [Π i, monoid (M i)]
+  [decidable_eq ι] [∀ i, decidable_eq (M i)]
+
+/-- `free_product M` is the free product, or categorical coproduct, of a family `M` of monoids.
+This assumes that the indexing type and each monoid all have decidable equality. -/
+@[derive decidable_eq, nolint unused_arguments]
+def free_product : Type* := { l : list (Σ i, { m : M i // m ≠ 1 }) // (l.map sigma.fst).chain' (≠) }
```

Comment:
Currently, the definition of `free_product` depends on the decidability instances. However, if I understand correctly, these are only needed to get decidable equality for the free product (and it is found automatically by type class inference, you don't need to prove this instance). Could you instead remove these decidability assumptions from the global context before the definition, and only add them when needed? For instance,
```

variables {ι : Type*} (M : Π i : ι, Type*) [Π i, monoid (M i)]

/-- `free_product M` is the free product, or categorical coproduct, of a family `M` of monoids.
This assumes that the indexing type and each monoid all have decidable equality. -/
@[nolint unused_arguments]
def free_product : Type* := { l : list (Σ i, { m : M i // m ≠ 1 }) // (l.map sigma.fst).chain' (≠) }

variables [decidable_eq ι] [∀ i, decidable_eq (M i)] {M}
```
works fine instead of these few lines. Of course, the linter will tell you where you really need these assumptions, and you should listen to what it says :-)

Reply:
I'm happy to make this change. I'll just point out that we need these decidability instances to define multiplication and the inclusions `of`. So without them there is not a lot you can do with this free product. Of course we could use classical instances, but then multiplication and `of` become noncomputable.

Reply:
From what I remember, decidability instances are much easier to handle in functions (like has_mul.mul) than they are if they end up in the type itself

Reply:
Ok, let's go with this then!